File: sdk/platform/Web/src/Landmark.cpp
#include <Posemesh/Landmark.hpp>
#include <emscripten/bind.h>
#include <memory>

using namespace emscripten;
using namespace psm;

namespace {
std::size_t hash(const Landmark& landmark) noexcept
{
    return std::hash<Landmark> {}(landmark);
}
}

EMSCRIPTEN_BINDINGS(Landmark)
{
    class_<Landmark>("Landmark")
        .smart_ptr<std::shared_ptr<Landmark>>("Landmark")
        .constructor(&std::make_shared<Landmark>)
        .constructor(&std::make_shared<Landmark, const Landmark&>)
        .function("duplicate()", &std::make_shared<Landmark, const Landmark&>, nonnull<ret_val>())
        .function("equals(landmark)", &Landmark::operator==)
        .function("hash()", &hash)
        .function("__getType()", &Landmark::getType)
        .function("__setType(type)", &Landmark::setType)
        .function("__getId()", &Landmark::getId)
        .function("__setId(id)", &Landmark::setId)
        .function("__getPosition()", &Landmark::getPosition)
        .function("__setPosition(position)", &Landmark::setPosition);
}

File: sdk/platform/Web/Landmark.js
posemeshModule.Landmark = null;

__internalPosemeshAPI.builderFunctions.push(function() {
    Object.defineProperty(__internalPosemesh.Landmark.prototype, 'type', {
        get: __internalPosemesh.Landmark.prototype.__getType,
        set: __internalPosemesh.Landmark.prototype.__setType,
        enumerable: true,
        configurable: false
    });

    Object.defineProperty(__internalPosemesh.Landmark.prototype, 'id', {
        get: __internalPosemesh.Landmark.prototype.__getId,
        set: __internalPosemesh.Landmark.prototype.__setId,
        enumerable: true,
        configurable: false
    });

    Object.defineProperty(__internalPosemesh.Landmark.prototype, 'position', {
        get: __internalPosemesh.Landmark.prototype.__getPosition,
        set: __internalPosemesh.Landmark.prototype.__setPosition,
        enumerable: true,
        configurable: false
    });

    posemeshModule.Landmark = __internalPosemesh.Landmark;
});
